server/src/main/java/org/opensearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -346,6 +374,67 @@ public String toString() {
         }
     }
 
+    @Override
+    protected TransportWriteAction.AsyncAfterWriteAction createAsyncAfterWriteAction(
+        IndexShard indexShard,
+        WriteRequest<?> request,
+        Translog.Location location,
+        TransportWriteAction.RespondingWriteResult respond,
+        Logger logger
+    ) {
+        return new AsyncAfterShardBulkAction(
+            indexShard,
+            request,
+            location,
+            respond,
+            logger,
+            latestSuccessfulLeaderCheckerTimeSupplier,
+            latestLeaderCheckThreshold
+        );
+    }
+
+    private static class AsyncAfterShardBulkAction extends AsyncAfterWriteAction {
+
+        private final IndexShard indexShard;
+        private final LongSupplier latestSuccessfulLeaderCheckerTimeSupplier;
+        private final TimeValue leaderCheckerTimeThreshold;
+
+        public AsyncAfterShardBulkAction(
+            final IndexShard indexShard,
+            final WriteRequest<?> request,
+            @Nullable final Translog.Location location,
+            final RespondingWriteResult respond,
+            final Logger logger,
+            final LongSupplier latestSuccessfulLeaderCheckerTimeSupplier,
+            final TimeValue leaderCheckerTimeThreshold
+        ) {
+            super(indexShard, request, location, respond, logger);
+            this.indexShard = indexShard;
+            this.latestSuccessfulLeaderCheckerTimeSupplier = latestSuccessfulLeaderCheckerTimeSupplier;
+            this.leaderCheckerTimeThreshold = leaderCheckerTimeThreshold;
+        }
+
+        @Override
+        protected void finishOnNoPendingOps() {
+            if (indexShard.isRemoteTranslogEnabled() && elapsedTimeSinceLatestSuccessfulLeaderCheckBeforeThreshold()) {
+                respond.onFailure(
+                    new LeaderCheckerBeforeThresholdException(
+                        "Latest successful leader checker epoch time is {} while threshold is {}",
+                        latestSuccessfulLeaderCheckerTimeSupplier.getAsLong(),
+                        leaderCheckerTimeThreshold
+                    )
+                );
+            } else {
+                super.finishOnNoPendingOps();
+            }
+        }
+
+        private boolean elapsedTimeSinceLatestSuccessfulLeaderCheckBeforeThreshold() {
```

Comment:
You should not rely on wall clock time. Instead rely on clock speeds using System.nanotime.

Reply:
Although `System.currentTimeMillis()` provides wall clock time, we are only interested in the difference between 2 instance of times. Since the comparison here, I am suspecting, would be in order of 100s of seconds, it did not make sense to use `System.nano()`. 

Also, as per java [documentation](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/System.html#currentTimeMillis()), nanoTime documentation says This method provides nanosecond precision, but not necessarily nanosecond resolution (that is, how frequently the value changes) - no guarantees are made except that the resolution is at least as good as that of [currentTimeMillis()](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/System.html#currentTimeMillis()).

In this usecase, afaik, nano time does seem to add much value. Also, do take a look at this - https://github.com/opensearch-project/OpenSearch/pull/6368. This was a fix for a flaky test where nanoTime was becoming problematic due to clock errors and time resolution.

Reply:
Read more about `System.currentTimeMillis()` and `System.nano()`.  The value returned by `System.currentTimeMillis()` in Java may change when the underlying machine's time is synchronized externally. While this is not the case for `System.nano()`. Considering this, it makes sense to use `System.nano()` time.